File: Wdtc/src/main/java/org/wdt/wdtc/ui/users/LittleskinWindow.java
package org.wdt.wdtc.ui.users;

import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.Pane;
import org.apache.log4j.Logger;
import org.wdt.wdtc.auth.UserList;
import org.wdt.wdtc.auth.yggdrasil.YggdrasilAccounts;
import org.wdt.wdtc.auth.yggdrasil.YggdrasilTextures;
import org.wdt.wdtc.manger.UrlManger;
import org.wdt.wdtc.ui.ErrorWindow;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;


public class LittleskinWindow {
    private static final Logger logmaker = WdtcLogger.getLogger(LittleskinWindow.class);

    public static void setLittleskinWin(Pane pane) {
        pane.getChildren().clear();
        Label littleskinTitle = new Label("Littleskin外置登录");
        littleskinTitle.setLayoutX(250.0);
        littleskinTitle.setLayoutY(69.0);
        Label username = new Label("用户名");
        username.setLayoutX(175.0);
        username.setLayoutY(107.0);
        TextField Inputusername = new TextField();
        Inputusername.setLayoutX(220.0);
        Inputusername.setLayoutY(103.0);
        Label powerwrod = new Label("密码:");
        powerwrod.setLayoutX(179.0);
        powerwrod.setLayoutY(135.0);
        TextField inputpowerword = new TextField();
        inputpowerword.setLayoutX(221.0);
        inputpowerword.setLayoutY(131.0);
        Label label = new Label();
        label.setLayoutX(220.0);
        label.setLayoutY(185.0);
        label.setPrefHeight(15.0);
        label.setPrefWidth(110.0);
        Button ok = new Button("登录");
        ok.setLayoutX(267.0);
        ok.setLayoutY(219.0);
        pane.getChildren().addAll(littleskinTitle, username, Inputusername, powerwrod, inputpowerword, label, ok);
        ok.setOnAction(event -> {
            String UserName = Inputusername.getText();
            String PowerWord = inputpowerword.getText();
            if (UserName.isEmpty() && PowerWord.isEmpty()) {
                label.setText("请输入用户名、密码");
                logmaker.warn("用户名、密码为空");
            } else {
                try {
                    YggdrasilAccounts yggdrasilAccounts = new YggdrasilAccounts(UrlManger.getLittleskinUrl(), Inputusername.getText(), inputpowerword.getText());
                    UserList.addUser(yggdrasilAccounts.getUser());
                    try {
                        YggdrasilTextures yggdrasilTextures = yggdrasilAccounts.getYggdrasilTextures();
                        yggdrasilTextures.DownloadUserSkin();
                        logmaker.info("Littleskin用户:" + Inputusername.getText() + "登陆成功!");
                        UserListPane.setUserList(pane);
                    } catch (IOException e) {
                        ErrorWindow.setErrorWin(e);
                    }
                } catch (IOException e) {
                    label.setText("用户名或密码错误");
                    logmaker.warn("用户名或密码错误", e);
                }
            }
        });

    }
}

File: Wdtc/src/main/kotlin/org/wdt/wdtc/ui/window/user/LittleskinWindow.kt
package org.wdt.wdtc.ui.window.user

import javafx.event.EventHandler
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.scene.layout.Pane
import kotlinx.coroutines.launch
import org.wdt.wdtc.core.auth.UsersList.Companion.saveChangeToFile
import org.wdt.wdtc.core.auth.currentUsersList
import org.wdt.wdtc.core.auth.yggdrasil.YggdrasilAccounts
import org.wdt.wdtc.core.manger.littleskinUrl
import org.wdt.wdtc.core.utils.ioCoroutineScope
import org.wdt.wdtc.core.utils.logmaker
import org.wdt.wdtc.core.utils.toCoroutineName
import org.wdt.wdtc.core.utils.warning
import java.io.IOException

object LittleskinWindow {
  fun setLittleskinWin(pane: Pane) {
    val littleskinTitle = Label("Littleskin外置登录").apply {
      layoutX = 250.0
      layoutY = 69.0
    }
    val username = Label("用户名").apply {
      layoutX = 175.0
      layoutY = 107.0
    }
    val inputusername = TextField().apply {
      layoutX = 220.0
      layoutY = 103.0
    }
    val powerWordTip = Label("密码:").apply {
      layoutX = 179.0
      layoutY = 135.0
    }
    val inputpowerword = TextField().apply {
      layoutX = 221.0
      layoutY = 131.0
    }
    val label = Label().apply {
      layoutX = 220.0
      layoutY = 185.0
      prefHeight = 15.0
      prefWidth = 110.0
    }
    val ok = Button("登录").apply {
      layoutX = 267.0
      layoutY = 219.0
      onAction = EventHandler {
        val userName = inputusername.text
        val powerWord = inputpowerword.text
        if (userName.isEmpty() && powerWord.isEmpty()) {
          label.text = "请输入用户名、密码"
        } else {
          try {
            loginUser(userName, powerWord)
            UserListPane.setUserList(pane)
          } catch (e: IOException) {
            label.text = "用户名或密码错误"
            logmaker.warning("用户名或密码错误", e)
          }
        }
      }
    }
    pane.children.run {
      clear()
      addAll(littleskinTitle, username, inputusername, powerWordTip, inputpowerword, label, ok)
    }
  }

  @Throws(IOException::class)
  private fun loginUser(userName: String, powerWord: String) {
    YggdrasilAccounts(
      littleskinUrl,
      userName, powerWord
    ).run {
      currentUsersList.apply {
        this@run.addToList()
      }.saveChangeToFile()

      ioCoroutineScope.launch("Login ${user.userName}".toCoroutineName()) {
        yggdrasilTextures.run {
          startDownloadUserSkin()
        }
      }
      logmaker.info("Login $userName Littleskin User")
    }
  }
}
